src/qforte/utils/moment_energy_corrections.py

```diff
@@ -0,0 +1,99 @@
+    ref = sum([b << i for i, b in enumerate(self._ref)])
+    # `& mask_alpha` gives the alpha component of a bitstring. `& mask_beta` does likewise.
+    mask_alpha = 0x5555555555555555
+    mask_beta = mask_alpha << 1
+    nalpha = sum(self._ref[0::2])
+    nbeta = sum(self._ref[1::2])
+
+    if not isinstance(self._max_moment_rank, int) or self._max_moment_rank < 0:
+        raise TypeError("The maximum moment rank must be a non-negative integer!")
+    elif self._max_moment_rank > nalpha + nbeta:
+        # WARNING: This step could be potetnially removed, if QForte allows for methods
+        #          that violate particle number symmetry
+        self._max_moment_rank = nalpha + nbeta
+
+    for I in range(1 << self._nqb):
+        alphas = [int(j) for j in bin(I & mask_alpha)[2:]]
+        betas = [int(j) for j in bin(I & mask_beta)[2:]]
+        # The following checks could be potentially removed, if QForte gets methods
+        # that violate particle number, Sz, and spatial symmetries
+        if sum(alphas) == nalpha and sum(betas) == nbeta:
+            if sq_op_find_symmetry(self._sys.orb_irreps_to_int,
+                                   [len(alphas) - i - 1 for i, x in enumerate(alphas) if x],
+                                   [len(betas) -i - 1 for i, x in enumerate(betas) if x]) == 0:
+                # Create the bitstring of created/annihilated orbitals
+                excit = bin(ref ^ I).replace("0b", "")
+                # Confirm excitation number is non-zero
+                if excit != "0":
+                    # Consider moments with rank <= self._max_moment_rank
+                    if int(excit.count('1')/2) <= self._max_moment_rank:
+                        occ_idx = [int(i) for i,j in enumerate(reversed(excit)) if int(j) == 1 and self._ref[i] == 1]
+                        unocc_idx = [int(i) for i,j in enumerate(reversed(excit)) if int(j) == 1 and self._ref[i] == 0]
+                        sq_op = qf.SQOperator()
+                        sq_op.add(+1.0, unocc_idx, occ_idx)
+                        sq_op.add(-1.0, occ_idx[::-1], unocc_idx[::-1])
+                        sq_op.simplify()
+                        self._mmcc_pool.add_term(0.0, sq_op)
+                        self._mpdenom.append(sum(self._orb_e[x] for x in occ_idx) - sum(self._orb_e[x] for x in unocc_idx))
+                        self._mmcc_excitation_indices.append(I)
+
+    for i in self._mmcc_pool.terms():
+        sq_op = i[1]
+        qc = qf.Computer(self._nqb)
+        qc.apply_circuit(self._Uprep)
+        # This could be replaced by set_bit?
+        qc.apply_operator(sq_op.jw_transform(self._qubit_excitations))
+        self._epstein_nesbet.append(qc.direct_op_exp_val(self._qb_ham))
+
+def compute_moment_energies(self):
+
+    # The moments, i.e., residuals, are estimated by measuring the "residual" state a la SPQE
+
+    if hasattr(self, '_dt'):
+        dt = self._dt
+    else:
+        # not sure about this
```

Comment:
What is "this"? The wisdom of this particular choice of `self._dt`? Having this attribute be `_dt` in the first place?

Reply:
In the case of SPQE, there exists an attribute called "_dt" and the user has the ability to change its value from the default of 0.001. The other algorithms, namely, UCCNPQE, UCCNVQE, and ADAPTVQE, do not have this attribute and creating it seems confusing and unnecessary. For such cases, the "dt" variable is hard-coded to 0.001 (the default for SPQE). The "not sure about this" refers to the hard-coding part. I am open to suggestions :-)

Reply:
Oh. I'd create a new variable rather than re-use `_dt`. Maybe call it `_moment_dt`. The two do (slightly) different things.

Reply:
Leave it flexible.